File: 20-actors/shionome/methods/test_kotoba.py
"""test_kotoba.py — 潮目 (shionome) kotoba Datom-log writer + content-addressed DAG. ADR-2606072200."""
from __future__ import annotations

import pathlib
import tempfile

import kotoba
from _edn import load_edn
from _t import run
from weave import weave

SEED = pathlib.Path(__file__).resolve().parents[1] / "data" / "seed-capital-flow-graph.kotoba.edn"


def _g():
    return weave(load_edn(SEED))


def test_graph_datoms_are_append_only():
    datoms = kotoba.graph_datoms(_g())
    assert datoms and all(d[0] == ":db/add" for d in datoms)   # no :db/retract (非終末論)


def test_tx_cid_deterministic():
    d = kotoba.graph_datoms(_g())
    assert kotoba.tx_cid(d, "") == kotoba.tx_cid(d, "")


def test_tx_cid_depends_on_prev():
    d = kotoba.graph_datoms(_g())
    assert kotoba.tx_cid(d, "") != kotoba.tx_cid(d, "bdeadbeef")


def test_append_and_read_roundtrip():
    with tempfile.TemporaryDirectory() as dr:
        log = pathlib.Path(dr) / "log.edn"
        tx = kotoba.make_tx(kotoba.graph_datoms(_g()), tx_id=1, as_of=20260607, prev_cid="")
        cid = kotoba.append_tx(tx, log)
        back = kotoba.read_log(log)
        assert len(back) == 1
        assert back[0][":tx/cid"] == cid


def test_chain_verifies_ok():
    with tempfile.TemporaryDirectory() as dr:
        log = pathlib.Path(dr) / "log.edn"
        prev = ""
        for i in range(1, 4):
            tx = kotoba.make_tx(kotoba.graph_datoms(_g()), tx_id=i, as_of=20260607 + i, prev_cid=prev)
            prev = kotoba.append_tx(tx, log)
        v = kotoba.verify_chain(log)
        assert v["ok"] is True and v["length"] == 3


def test_head_cid_is_last():
    with tempfile.TemporaryDirectory() as dr:
        log = pathlib.Path(dr) / "log.edn"
        last = ""
        for i in range(1, 3):
            tx = kotoba.make_tx(kotoba.graph_datoms(_g()), tx_id=i, as_of=20260607 + i,
                                prev_cid=kotoba.head_cid(log))
            last = kotoba.append_tx(tx, log)
        assert kotoba.head_cid(log) == last


def test_tamper_breaks_chain():
    with tempfile.TemporaryDirectory() as dr:
        log = pathlib.Path(dr) / "log.edn"
        for i in range(1, 3):
            tx = kotoba.make_tx(kotoba.graph_datoms(_g()), tx_id=i, as_of=20260607 + i,
                                prev_cid=kotoba.head_cid(log))
            kotoba.append_tx(tx, log)
        # tamper: flip a byte in the first tx's value
        lines = log.read_text(encoding="utf-8").splitlines()
        lines[1] = lines[1].replace("US equities", "TAMPERED")
        log.write_text("\n".join(lines) + "\n", encoding="utf-8")
        assert kotoba.verify_chain(log)["ok"] is False


def test_post_datoms_status_dry_run():
    posts = [{":post/subject": "netflow", ":post/status": ":dry-run", ":post/body": "x"}]
    datoms = kotoba.post_datoms(posts)
    statuses = [d[3] for d in datoms if d[2] == ":post/status"]
    assert statuses == [":dry-run"]


def test_post_body_with_newlines_roundtrips():
    # the bug that broke the DAG: a post body has \n\n; it must survive write→read
    with tempfile.TemporaryDirectory() as dr:
        log = pathlib.Path(dr) / "log.edn"
        posts = [{":post/subject": "regime", ":post/status": ":dry-run",
                  ":post/body": "line one\n\nline two with 日本語"}]
        tx = kotoba.make_tx(kotoba.post_datoms(posts), tx_id=1, as_of=20260607, prev_cid="")
        kotoba.append_tx(tx, log)
        assert kotoba.verify_chain(log)["ok"] is True


if __name__ == "__main__":
    run("kotoba", [(n, f) for n, f in sorted(globals().items())
                   if n.startswith("test_") and callable(f)])

File: 20-actors/shionome/methods/test_kotoba.clj
#!/usr/bin/env bb
;; test_kotoba.clj — shionome kotoba Datom-log writer + content-addressed DAG. ADR-2606072200.
;; Clojure port of test_kotoba.py (9 tests).
(ns shionome.methods.test-kotoba
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [clojure.java.io :as io]
            [shionome.methods.kotoba :as kotoba]
            [shionome.methods.edn :as sedn]
            [shionome.methods.weave :as weave]))

(def ^:private SEED
  (-> *file* io/file .getAbsoluteFile .getParentFile .getParentFile
      (io/file "data" "seed-capital-flow-graph.kotoba.edn")))

(defn- g []
  (weave/weave (sedn/load-edn SEED)))

(defn- tmp-log []
  (doto (java.io.File/createTempFile "shionome-kotoba-test-" ".edn")
    .deleteOnExit))

;; ── tests ─────────────────────────────────────────────────────────────────────

(deftest test-graph-datoms-are-append-only
  (let [datoms (kotoba/graph-datoms (g))]
    (is (seq datoms) "datoms is non-empty")
    (is (every? #(= ":db/add" (first %)) datoms)
        "no :db/retract (非終末論)")))

(deftest test-tx-cid-deterministic
  (let [d (kotoba/graph-datoms (g))]
    (is (= (kotoba/tx-cid d "") (kotoba/tx-cid d ""))
        "tx-cid is deterministic for same datoms + prev")))

(deftest test-tx-cid-depends-on-prev
  (let [d (kotoba/graph-datoms (g))]
    (is (not= (kotoba/tx-cid d "") (kotoba/tx-cid d "bdeadbeef"))
        "tx-cid changes when prev-cid changes")))

(deftest test-append-and-read-roundtrip
  (let [log (tmp-log)
        log-path (.getAbsolutePath log)
        tx  (kotoba/make-tx (kotoba/graph-datoms (g))
                            :tx-id 1 :as-of 20260607 :prev-cid "")
        cid (kotoba/append-tx tx log-path)
        back (kotoba/read-log log-path)]
    (is (= 1 (count back)) "one tx read back")
    (is (= cid (get (first back) ":tx/cid"))
        "CID round-trips through EDN write+read")))

(deftest test-chain-verifies-ok
  (let [log (tmp-log)
        log-path (.getAbsolutePath log)]
    (loop [prev "" i 1]
      (when (<= i 3)
        (let [tx (kotoba/make-tx (kotoba/graph-datoms (g))
                                 :tx-id i :as-of (+ 20260607 i) :prev-cid prev)
              cid (kotoba/append-tx tx log-path)]
          (recur cid (inc i)))))
    (let [v (kotoba/verify-chain log-path)]
      (is (true? (:ok v)) "chain verifies ok")
      (is (= 3 (:length v)) "length is 3"))))

(deftest test-head-cid-is-last
  (let [log (tmp-log)
        log-path (.getAbsolutePath log)]
    (let [last-cid
          (loop [i 1 last-cid ""]
            (if (> i 2)
              last-cid
              (let [tx  (kotoba/make-tx (kotoba/graph-datoms (g))
                                        :tx-id i :as-of (+ 20260607 i)
                                        :prev-cid (kotoba/head-cid log-path))
                    cid (kotoba/append-tx tx log-path)]
                (recur (inc i) cid))))]
      (is (= last-cid (kotoba/head-cid log-path))
          "head-cid is the last appended CID"))))

(deftest test-tamper-breaks-chain
  (let [log (tmp-log)
        log-path (.getAbsolutePath log)]
    ;; append 2 txs
    (loop [i 1]
      (when (<= i 2)
        (let [tx (kotoba/make-tx (kotoba/graph-datoms (g))
                                 :tx-id i :as-of (+ 20260607 i)
                                 :prev-cid (kotoba/head-cid log-path))]
          (kotoba/append-tx tx log-path)
          (recur (inc i)))))
    ;; tamper: flip "US equities" → "TAMPERED" in the first tx line
    (let [text  (slurp log-path :encoding "UTF-8")
          lines (clojure.string/split-lines text)
          ;; lines[0] = header comment, lines[1] = first tx
          tampered-lines (assoc (vec lines) 1
                                (clojure.string/replace (nth lines 1) "US equities" "TAMPERED"))]
      (spit log-path (str (clojure.string/join "\n" tampered-lines) "\n") :encoding "UTF-8"))
    (is (false? (:ok (kotoba/verify-chain log-path)))
        "tampered chain fails verification")))

(deftest test-post-datoms-status-dry-run
  (let [posts   [{":post/subject" "netflow"
                  ":post/status"  ":dry-run"
                  ":post/body"    "x"}]
        datoms  (kotoba/post-datoms posts)
        statuses (mapv #(nth % 3) (filter #(= ":post/status" (nth % 2)) datoms))]
    (is (= [":dry-run"] statuses)
        "post status is :dry-run")))

(deftest test-post-body-with-newlines-roundtrips
  ;; the bug that broke the DAG: a post body has \n\n; it must survive write→read
  (let [log (tmp-log)
        log-path (.getAbsolutePath log)
        posts [{":post/subject" "regime"
                ":post/status"  ":dry-run"
                ":post/body"    "line one\n\nline two with 日本語"}]
        tx  (kotoba/make-tx (kotoba/post-datoms posts)
                            :tx-id 1 :as-of 20260607 :prev-cid "")]
    (kotoba/append-tx tx log-path)
    (is (true? (:ok (kotoba/verify-chain log-path)))
        "chain with newline body verifies ok")))

;; ── runner ────────────────────────────────────────────────────────────────────

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'shionome.methods.test-kotoba)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
